src/storage/mod.rs

```diff
@@ -0,0 +1,152 @@
+use thiserror::Error;
+
+mod nats_kv;
+mod state;
+
+use async_trait::async_trait;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use state::LatticeState;
+
+pub use nats_kv::{NatsAuthConfig, NatsKvStorageConfig, NatsKvStorageEngine};
+pub use state::{
+    Actor, ActorBuilder, Claim, ClaimBuilder, Host, HostBuilder, LatticeParameters,
+    LatticeStateBuilder, Provider, ProviderBuilder,
+};
+
+/////////////
+// Engines //
+/////////////
+
+/// Pluggable engines can be asked for their metadata
+/// to enable generic (possibly dyn) usage
+#[async_trait]
+pub trait EngineMetadata {
+    /// Get the name of the lattice storage
+    async fn name() -> String;
+
+    /// Get the version of the lattice storage
+    async fn version() -> Version;
+}
+
+/////////////////////
+// Storage Engines //
+/////////////////////
+
+/// Errors that are related to storage engines
+#[derive(Debug, Error)]
+pub enum StorageEngineError {
+    /// When the underlying engine throws an error
+    #[error("Storage engine encountered an error: {0}")]
+    Engine(String),
+
+    #[error("Storage engine [{0}] not connected")]
+    NotConnected(String),
```

Comment:
Because we don't have a connect method any more, do we need this error? If so, should we just not have a top level error and instead allow a store to specify its own error type?